src/import/duplicate_detect.rs

```diff
@@ -56,110 +55,126 @@ impl<S: Snapshot> DuplicateDetector<S> {
-                                    .unwrap_or(0);
-                                ret.push(self.make_kv_pair(
-                                    &last_key,
-                                    Some(last_write),
-                                    last_commit_ts,
-                                )?);
-                            }
-                        }
-                        let w = if self.key_only {
-                            None
-                        } else {
-                            batch_size += write.short_value.as_ref().map(|v| v.len()).unwrap_or(0);
-                            Some(write)
-                        };
-                        ret.push(self.make_kv_pair(&user_key, w, commit_ts)?);
-                        first_duplicate = false;
-                        if commit_ts <= self.min_commit_ts {
-                            has_delete_key_before_import = true;
-                        }
-                    } else {
-                        if ret.len() >= MAX_SCAN_BATCH_COUNT || batch_size >= MAX_SCAN_BATCH_SIZE {
-                            return Ok(Some(ret));
-                        }
-                        first_duplicate = true;
-                        has_delete_key_before_import = false;
-                        if commit_ts > self.min_commit_ts {
-                            has_put_key_before_import = false;
-                        } else {
-                            has_put_key_before_import = true;
-                        }
-                        if !self.key_only {
-                            last_value.clear();
-                            last_value.extend_from_slice(value);
-                        }
+            } else {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    fn skip_lock_and_rollback(&mut self, start_key: &[u8]) -> Result<Option<Write>> {
+        while self.iter.valid().map_err(from_kv_error)? {
+            let (current_key, _) = Key::split_on_ts_for(self.iter.key())?;
+            if !current_key.eq(start_key) {
+                return Ok(None);
+            } else {
+                let write = WriteRef::parse(self.iter.value()).map_err(from_txn_types_error)?;
+                match write.write_type {
+                    WriteType::Put | WriteType::Delete => return Ok(Some(write.to_owned())),
+                    WriteType::Lock | WriteType::Rollback => {
+                        self.iter.next().map_err(from_kv_error)?;
                     }
                 }
             }
+        }
+        Ok(None)
+    }
 
-            last_key.clear();
-            last_key.extend_from_slice(user_key);
-            last_commit_ts = commit_ts;
+    fn skip_all_version(&mut self, start_key: &[u8]) -> Result<()> {
+        while self.iter.valid().map_err(from_kv_error)? {
+            let (current_key, _) = Key::split_on_ts_for(self.iter.key())?;
+            if !current_key.eq(start_key) {
```

Comment:
Why not use the `==` and `!=` operators when comparing keys?

Reply:
Because some language `==` just compare pointer address rather than the content of struct.

Reply:
@Little-Wallace I don't see how languages other than Rust matter here.